### ISSUE TITLE
Add a monotonic clock to the framework

### ISSUE THREAD

**Issue by GSPP** [dotnet/corefx#3249]
**Problem:** Many BCL components use `DateTime.UtcNow` to compute timeouts and time deltas. When you run a "Find Usages" on `DateTime.UtcNow` in .NET Framework assemblies a lot of usage sites come to light. For example, in ADO.NET, transactions, caches, WebRequest, remoting, WCF, SignalR, security.

**All of these usages are bugs** because the current system time can change significantly forwards and backwards at any time. This means that in case the system clock is changed many .NET components will spuriously fail.

Common symptoms would be timeouts firing to early or too late (possibly never). Perf counters also might show too high or negative values.

User code also has the same problems.

**Solution:** Add some kind of monotonic clock to the BCL. The main property of such a clock would be that its time advances linearly. It should never jump when the system clock changes. It should not exhibit split second jumps. It should behave reasonably in case the system sleeps or hibernates or in case the VM is paused.

This monotonic clock class should be fast and convenient to use so that it is a no-brainer to switch all `DateTime.UtcNow` usages over to the new model.

Maybe we can add `Environment.TickCount64` as well. `Environment.TickCount64` alone would not be good enough because it is awkward to use.

I'm not sure if `Stopwatch` would cover all these requirements. It is a reference type which might be too heavy for hot code path in the BCL.

**Comment by stephentoub** [comment]
> It is a reference type which might be too heavy for hot code path in the BCL.

Just responding to this one point, Stopwatch provides the static GetTimestamp() method, so if the timing support provided by Stopwatch is what you want, you can get at the same mechanism without instantiating a Stopwatch instance.

**Comment by CodesInChaos** [comment]
static MonotonicTime ForDateTime(DateTime utcNow);
    DateTime AsDateTime();

How would these work? If these function existed, it'd mean that `MonotonicTime` is equivalent to `DateTime` and `UtcNow`, which clearly contradicts the requirements of a monotonic clock.

**Comment by KrzysztofCwalina** [comment]
I can imagine this implementation (minus synchronization):

struct MonotonicTime {
   static long maxTicks;
   static MonotonicTime ForDateTime(DateTime utcNow){
      if(utcNow.Ticks > maxTicks) {
          maxTicks = utcNow.Ticks;
      }
      return new MonotonicTime(maxTicks);
   }
}

**Comment by GSPP** [comment]
@CodesInChaos true, I have edited those away.

**Comment by GSPP** [comment]
@KrzysztofCwalina that's not good enough because it can cause time to stand still for arbitrary periods of time.

I think `MonotonicTime` should be based on `Environment.TickCount64` or `Stopwatch.GetTimestamp` or whatever it is called. The latter is nice because it has high precision. Could be nice for logging when you can tell action apart that happened within microseconds. I always found the limited 15ms precision of DateTime.UtcNow and `Environment.TickCount` quite ugly.

**Comment by KrzysztofCwalina** [comment]
@GSPP, I thought monotonic means never decreasing. Also, I am not sure what the semantics/implementation could be if you wanted to always increasing. Given the system clock is lower resolution than processor clock, it's not clear how it would be implemented.

**Comment by svick** [comment]
@KrzysztofCwalina The way I understand it, monotonic *does* mean never decreasing, but that's not the only requirement. It's okay for the value to not increase for, say, 366 ns (that's what the resolution of `Stopwatch` seems to be for me). It's definitely not okay for it to not increase for an arbitrary amount of time.

Also, your implementation is not really monotonic (at least not with this interface). Consider something like:

```c#
var dt1 = DateTime.UtcNow;
Thread.Sleep(aWhile);
var dt2 = DateTime.UtcNow;
Thread.Sleep(someMore);
var dt3 = DateTime.UtcNow;
    
var monotonic2 = MonotonicTime.ForDateTime(dt2);
var monotonic3 = MonotonicTime.ForDateTime(dt3);
var monotonic1 = MonotonicTime.ForDateTime(dt1);
```

Here, `monotonic1 > monotonic2`, which is wrong.

**Comment by GSPP** [comment]
@svick correct, it's supposed to be like `TickCount`. If it's not suitable for timeouts or relative time measurement it's no good. The more resolution and the cheaper to compute the better.

**Comment by KrzysztofCwalina** [comment]
@GSPP, @svick, Can you then propose an implementation? I am not sure how to get the semantics you want.

**Comment by GSPP** [comment]
@KrzysztofCwalina I think Stopwatch timestamps should work. According to search engines they even continue advancing when the system is hibernated.

**Comment by KrzysztofCwalina** [comment]
@GSPP, ah, ok. I got confused. I thought you did not want to use Stopwatch/QueryPerformanceCounter for so reason.

**Comment by scalablecory** [comment]
One note is that `Stopwatch` actually falls back on `DateTime.UtcNow.Ticks` if performance counters are not available, so even then it's not a robust solution.

**Comment by GSPP** [comment]
@scalablecory good point but maybe there are no supported platforms anymore where that fallback is necessary. If this is rare enough one could even make the monotonic time struct throw in case good stopwatch support is not there.

**Comment by mj1856** [comment]
Just stumbled across this issue and I'll add:

- `Stopwatch` is indeed the most appropriate timer for profiling code (deltas). It uses `QueryPerformanceCounter`, which you can read about [here](https://msdn.microsoft.com/en-us/library/windows/desktop/dn553408.aspx).

- `Stopwatch.GetTimestamp()` is indeed the way to go when profiling *large quantities* of code.  It prevents the additional heap allocations and GC you'd have with lots of `Stopwatch` instances.  Be sure to take the `Frequency` property into account when computing the result.

- The idea of providing UTC time with higher precision is not new, but neither `Stopwatch` or `DateTime.UtcNow` are the most appropriate sources for this (nor are binding them together).
  - Windows actually provides a native function for this [`GetSystemTimePreciseAsFileTime`](https://msdn.microsoft.com/library/windows/desktop/hh706895.aspx).  I can see a .Net wrapper around that being useful for certain scenarios.
  - I'm not 100% sure what the Linux equivalent is (probably some flavor of [`clock_gettime`](http://linux.die.net/man/3/clock_gettime)).

- The remarks in the MSDN for the `GetSystemTimePreciseAsFileTime` summarize the usage scenarios quite well:

  > **Note**  This function is best suited for high-resolution time-of-day measurements, or time stamps that are synchronized to UTC. For high-resolution interval measurements, use `QueryPerformanceCounter` ...

  And as mentioned previously, QPC is already used by `Stopwatch`.

WRT the original scenarios mentioned, many are indeed just fine with `UtcNow`.  For example, expiring a cache in an hour is not an operation that requires high precision.  In fact, trying to provide such high precision can be extremely difficult when you consider distributed environments.  Eventually you reach the limitations of NTP and w32tm, and would have to have [specialized hardware](https://www.meinbergglobal.com/english/products/) to really obtain such accurate time coordination across machines.

**Comment by GSPP** [comment]
`GetSystemTimePreciseAsFileTime` seems like a highly useful API. Can we switch `DateTime.UtcNow` over to it? Needs a benchmark. Otherwise this seems like a pure win.

But it's not monotonic because the clock can be changed.

The idea of a monotonic clock does not include providing "legible" times as DateTime.UtcNow provides it. This notion is *incompatible* with having a monotonic clock. A monotonic clock can only meaningfully provide time deltas. Any absolute time stamp would be a meaningless number with a system dependent unit exactly like the `Stopwatch` timestamps.

This ticket is *not* about high precision so much. It's about a clock source that cannot jump arbitrarily like the system clock.

**Comment by scalablecory** [comment]
Re: `GetSystemTimePreciseAsFileTime`, given that it has an identical API to `GetSystemTimeAsFileTime`, I suspect it has some performance implications so we'd be unnecessarily slowing a lot of apps down for precision they don't need if we used it for `DateTime.UtcNow`.

.NET does already expose all the monotonic clocks most would ever need, between `Stopwatch` (which should only be used if `IsHighResolution` is `true`) and `Environment.TickCount`.

**Comment by GSPP** [comment]
The API is new to Windows 8. It did not exist back then. I have created a ticket to split off that discussion. It has nothing to do with monotonic time.

`Environment.TickCount` is unusable because the code will break after 30 days of running. `Stopwatch` appears to be a suitable source of monotonic time, as discussed above. My API request stands. It's a convenience API.

https://github.com/dotnet/coreclr/issues/5061

**Comment by RobThree** [comment]
A stricly increasing (e.g. non-decreasing) clock I know of (and that is, 'as a bonus'?, high precision) is [QueryPerformanceCounter](https://msdn.microsoft.com/en-us/library/windows/desktop/dn553408(v=vs.85).aspx).

> Is the performance counter monotonic (non-decreasing)?
> Yes

It can 'leap' forward AFAIK but never back.

**Comment by scalablecory** [comment]
@GSPP proper long-term use of `Environment.TickCount` would be to query it at least every 30 days and maintain a total.

Of course a `TickCount64` would be best, and could work as far back as XP SP2 (i know the docs for `GetTickCount64` say Vista, but `GetTickCount` actually returns a 64-bit value in XP SP2).

I think ideally Stopwatch.GetTimestamp would just be updated to fall back on `GetTickCount64`, so it would be the single robust monotonic solution.

**Comment by GSPP** [comment]
Right, Stopwatch should just fall back to the 64 bit counter. This is another change request.

> proper long-term use of Environment.TickCount would be to query it at least every 30 days and maintain a total.

To nitpick, your code is not guaranteed to be scheduled once every 30 days. It could be a suspended VM or a sleeping physical machine. Since the system clock can change solutions based on detecting wrapping that way also fail.

To summarize:

1. The framework timeouts should be fixed. (The TPL bugs apparently are already fixed. I reported these a while ago.)
2. Add `Environment.TickCount64`.
3. Provide a convenient, high-performance monotonic clock.
4. Make Stopwatch fall back onto the 64 bit TickCount counter. Right now a fallback Stopwatch can jump which is really bad.
5. [Use GetSystemTimePreciseAsFileTime for DateTime.UtcNow](https://github.com/dotnet/coreclr/issues/5061)

**Comment by scalablecory** [comment]
It is worth noting that I submitted a bug to MS ~6 years ago and was denied with "wont fix" regarding `Stopwatch` using `DateTime` due to compatibility reasons. It sure would be nice if their reasoning changed. https://connect.microsoft.com/VisualStudio/feedback/details/622002/low-resolution-stopwatch-behaves-incorrectly

**Comment by mj1856** [comment]
@RobThree - My understanding is that the issues described in [KB274323](https://web.archive.org/web/20070223010511/http://support.microsoft.com/kb/274323), and in [this blog post](http://kristofverbiest.blogspot.co.uk/2008/10/beware-of-stopwatch.html) relating to QPC being unreliable are no longer accurate because 1) The hardware that is known to exhibit this bug is very old and no longer in common usage, and 2) There were some workarounds added to detect this hardware bug and use an alternate timing source.  See the notes under TSC Register at the bottom of the [Acquiring hi-res timestamps](https://msdn.microsoft.com/en-us/library/windows/desktop/dn553408.aspx) article.

**Comment by mj1856** [comment]
@GSPP - Agreed with 1, 2 & 4.  Not sure what else is necessary for 3 that Stopwatch doesn't provide.  5 we'll talk about in the other item. :smile:

**Comment by dasMulli** [comment]
> Provide a convenient, high-performance monotonic clock.

An additional benefit of a monotonic clock could be correct handling of leap seconds, which `DateTime` currently can't handle. So any code that currently takes two results of `DateTime.UtcNow` and compares them may be off by up to a second every few years which is unfortunate for checking short timeouts.

**Comment by Clockwork-Muse** [comment]
> An additional benefit of a monotonic clock could be correct handling of leap seconds, which `DateTime` currently can't handle. So any code that currently takes two results of `DateTime.UtcNow` and compares them may be off by up to a second every few years which is unfortunate for checking short timeouts.

Computers/Programs have a bad time with leap seconds in general.  A monotonic clock doesn't actually handle leap seconds so much as it wouldn't notice there was anything different.  If you tried to translate it back to a readable date/time value, you'd still get translation errors (unless there was some sort of mapping).  `DateTime` has no ability to deal with leap seconds, though, so unless we add a new full date/time library, I'd recommend just ignoring it.

**Comment by dasMulli** [comment]
> If you tried to translate it back to a readable date/time value, you'd still get translation errors

Exactly! The proposed API by @GSPP doesn't include a conversion to/from DateTime and so the type system will prevent you from doing just that. So comparing two instances of `MonotonicTime` could tell you how much metric time elapsed between them independent of how the system clock setting or even UTC behaves. If i'm not mistaken, resetting the clock is how leap seconds are handled on many systems and that is exactly the scenario @GSPP started with - just with a second instead of a year.
And yes, i might be the only one caring about leap seconds..

**Comment by Clockwork-Muse** [comment]
> The proposed API by @GSPP doesn't include a conversion to/from DateTime and so the type system will prevent you from doing just that. 

Heh.  `StartTime + Elapsed` still has its own problems, though.  We should definitely make use of this internally, and expose it for these types of scenarios, but it's going to need some warnings about blindly using the deltas it generates.  When you log, for example, you're almost certainly going to want to take current (UTC) time, and ignore this.

> If i'm not mistaken, resetting the clock is how leap seconds are handled on many systems 

Not quite sure you're going with this.  Ignoring NTP syncing problems, OS/runtime clocks do any or all of the following:
 - blow up ("second == 60!?")
 - smear the second across the entire minute/day (Java - at least for the human-friendly classes, google?, a few others)
 - repeat the second
 - monotonically increase the time to the next minute (ignore the leap second)

**Comment by AlexGhiondea** [comment]
@KrzysztofCwalina what are the next steps here? Are we waiting for a formal API proposal?

**Comment by GSPP** [comment]
I (kind of) made an API proposal. Regardless of whether this ends up being a community contribution this should make it into the framework. The point of this ticket was to make the team aware of a need and a solution. I did not try to apply for implementing this.

**Comment by AlexGhiondea** [comment]
@GSPP thanks for pointing out the problem and offering the start of the API shape!

I think we need to think about this some more and come up with the final API proposal that we can submit for review.

Anyone interested in picking this up?

**Comment by mj1856** [comment]
Well, if it's going to be a new API, it should be part of `System.Time` in CoreFXLab, IMHO.  So I guess that means me. 😉 

I'm still not clear exactly on the proposed API.  Sure - subtracting two instances of `MonotonicTime` should give a `TimeSpan` that is as accurate and precise as possible with regards to elapsed time.  But what does a *single* instance of `MonotonicTime` actually represent?  How should it be displayed?  Should there be any conversions when interacting with `DateTime` or `DateTimeOffset`?  Does it need a `.Now` or similar property or method that is somehow related to the system clock?  Or does it just track ticks?

Ultimately, are we just putting a value-type wrapper around Stopwatch ticks?  Or are we somehow tying it back to UTC?

If your intention was to track something without leap seconds, like [TAI](https://en.wikipedia.org/wiki/International_Atomic_Time), we would actually need to maintain a table of leap seconds, updating the nuget package whenever new leap seconds were announced.  This is possible, but how useful is that?  IDK.

**Comment by GSPP** [comment]
@mj1856 in my mind there is no relation to the wall clock at all. This class would never be exposed to the user.

Any "debug" formatting would suffice. Strawman: "12345678 ticks (likely 2017-01-23 18:32:12.345 local time)". This *does* have a relationship to the wallclock but it's only for the developers convenience.

I don't see the need to convert to wallclock time except as a debugging aid. This is conceptually impossible.

It is possible to approximately convert to wallclock time by storing the current tick count and current DateTime.UtcNow in a static field before the first instance is created (in the cctor). That way an UTC date can be obtained relative to the time the cctor ran. Again, this is a debugging aid. The method could be called `MonotonicTime.GetApproximateDateTimeUTC/Local()`.

I think this would indeed just be a value-type wrapper around Stopwatch ticks, yes. This is a convenience API for a few very common scenarios outlined in the opening post.

Regarding `Now`, I propose `static MonotonicTime Current { get; }`.

**Comment by mj1856** [comment]
@GSPP  - Oh, so you're just proposing a more strongly typed tick?  I suppose that would help with people forgetting to take frequency into account with StopWatch ticks, but other than that, is it just an ease of use issue?  I mean, we do have `StopWatch.GetTimestamp` already.

**Comment by GSPP** [comment]
@mj1856 yes, strongly typed ticks to improve usability to the point where people actually do this instead of hacking it with `DateTime.UtcNow` (which is broken) and `Environment.TickCount` which is not handy and overflows (also broken therefore).
